### PR TITLE
[SPARK-41297] [CONNECT] [PYTHON] Support String Expressions in filter.

### PR DESCRIPTION
**Pull request by grundprinzip**
### What changes were proposed in this pull request?
This patch extends the functionality of the DataFrame in Spark Connect by support passing string based SQL expressions to the filter. The expressions are passed as is to the server during evaluation. If the expression is not boolean an Analysis error is thrown.

```
spark.range(100).filter("id < 10").collect()
```

### Why are the changes needed?
Compatibility

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
UT

### CODE REVIEW

File: python/pyspark/sql/connect/dataframe.py [hunk 1]
```diff
@@ -325,10 +325,14 @@ def drop(self, *cols: "ColumnOrName") -> "DataFrame":
             session=self._session,
         )
 
-    def filter(self, condition: Expression) -> "DataFrame":
-        return DataFrame.withPlan(
-            plan.Filter(child=self._plan, filter=condition), session=self._session
-        )
+    def filter(self, condition: Union[Expression, str]) -> "DataFrame":
+        # The expression can be a SQL string expression.
```
Comment (amaliujia) [inline python/pyspark/sql/connect/dataframe.py:329]: Nit:

I think this comment is a bit redundant as the code below is self-explained.

**Review by amaliujia — COMMENTED** [review]
LGTM with a minor comment

**Review by HyukjinKwon — APPROVED** [review]

**Comment by zhengruifeng** [comment]
python linter failed

**Comment by grundprinzip** [comment]
Fixed the linter issue

**Review by hvanhovell — APPROVED** [review]
LGTM

File: python/pyspark/sql/tests/connect/test_connect_plan_only.py [hunk 2]
```diff
@@ -70,6 +70,14 @@ def test_filter(self):
         self.assertEqual(plan.root.filter.condition.unresolved_function.parts, [">"])
         self.assertEqual(len(plan.root.filter.condition.unresolved_function.arguments), 2)
 
+    def test_filter_with_string_expr(self):
+        """SPARK-41297: filter supports SQL expression"""
```
Comment (HyukjinKwon) [inline python/pyspark/sql/tests/connect/test_connect_plan_only.py:74]: just dropping a comment but it should be:

```
# SPARK-41297: filter supports SQL expression
```

per https://spark.apache.org/contributing.html

**Pull request closed without merging**